CMakeLists.txt

```diff
@@ -1,87 +1,89 @@
-
-# Create the top level simdjson library (must be done at this level to use both src/ and include/
-# directories) and tools
-#
-add_subdirectory(include)
-add_subdirectory(src)
-add_subdirectory(windows)
-if(NOT(SIMDJSON_JUST_LIBRARY))
-  add_subdirectory(dependencies) ## This needs to be before tools because of cxxopts
-  add_subdirectory(tools)  ## This needs to be before tests because of cxxopts
-  add_subdirectory(singleheader)
-endif()
-install(FILES singleheader/simdjson.h DESTINATION ${CMAKE_INSTALL_INCLUDEDIR})
-
-#
-# Compile tools / tests / benchmarks
-#
-if(NOT(SIMDJSON_JUST_LIBRARY))
-  add_subdirectory(tests)
-  add_subdirectory(examples)
-  add_subdirectory(benchmark)
-  add_subdirectory(fuzz)
-endif()
+# ---- Warning guard ----
 
-#
-# Source files should be just ASCII
-#
-find_program(FIND find)
-find_program(FILE file)
-find_program(GREP grep)
-if((FIND) AND (FILE) AND (GREP))
-    add_test(
-      NAME "just_ascii"
-      COMMAND sh -c "${FIND}  include src windows tools singleheader tests examples benchmark -path benchmark/checkperf-reference -prune -name '*.h'  -o -name '*.cpp' -type f  -exec ${FILE} '{}' \; |${GREP} -v ASCII || exit 0  && exit 1"
-      WORKING_DIRECTORY ${CMAKE_CURRENT_SOURCE_DIR}
-    )
+# Protect dependents from this project's warnings if the guard isn't disabled
+set(simdjson_warning_guard "SYSTEM")
+if(simdjson_INCLUDE_WITHOUT_SYSTEM)
+  set(simdjson_warning_guard "")
 endif()
 
+# ---- Declare library ----
+
+include(cmake/sources.cmake)
+
+add_library(simdjson ${sources})
+add_library(simdjson::simdjson ALIAS simdjson)
+
+set_target_properties(simdjson PROPERTIES
+        VERSION ${SIMDJSON_LIB_VERSION}
+        SOVERSION ${SIMDJSON_LIB_SOVERSION}
+        WINDOWS_EXPORT_ALL_SYMBOLS YES)
+
+target_compile_features(simdjson PUBLIC cxx_std_11)
+target_include_directories(simdjson
+        ${simdjson_warning_guard}
+        PUBLIC
+        "$<BUILD_INTERFACE:${PROJECT_SOURCE_DIR}/include>")
+
+include(cmake/implementations.cmake)
+include(cmake/options.cmake)
+
+# ---- Install ----
+
+include(CMakePackageConfigHelpers)
+include(GNUInstallDirs)
+
+set(simdjson_directory "simdjson-${PROJECT_VERSION}")
+set(simdjson_include_directory
+        "${CMAKE_INSTALL_INCLUDEDIR}/${simdjson_directory}")
+
+install(TARGETS simdjson
+        EXPORT simdjsonTargets
+        RUNTIME #
+        DESTINATION "${CMAKE_INSTALL_BINDIR}"
+        COMPONENT simdjson_Runtime
+        LIBRARY #
+        DESTINATION "${CMAKE_INSTALL_LIBDIR}"
+        COMPONENT simdjson_Runtime
+        NAMELINK_COMPONENT simdjson_Development
+        ARCHIVE #
+        DESTINATION "${CMAKE_INSTALL_LIBDIR}"
+        COMPONENT simdjson_Development
+        INCLUDES #
+        DESTINATION "${simdjson_include_directory}")
+
+set(simdjson_install_cmakedir
+        "${CMAKE_INSTALL_LIBDIR}/cmake/${simdjson_directory}")
+
+write_basic_package_version_file(
+        simdjsonConfigVersion.cmake
+        VERSION ${PROJECT_VERSION}
```

Comment:
Should `PROJECT_VERSION` be used here or the fake `SIMDJSON_LIB_VERSION`?

This argument here affects people importing the library with version requirements:
```cmake
find_package(simdjson 4) # will match 4.x.x.x if COMPATIBILITY is SameMajorVersion
```

Reply:
Please do not change anything anything having to do with versions. We semantic versioning as well as library versioning, and a so version, and we have a scripted release process that updates automatically the files and CMake files, keeping everything in sync. All lines where versions are specify need to remain unchanged.

If you want to change how simdjson is versioned, please open a distinct issue, and let us discuss it. We won't change versioning without very deliberate and very thorough discussions.

We have a system that works currently, it meets the needs of our downstream users. We have a good release process. We don't want to break any of this.

How versioning works on the different systems we support is a complex set of conventions.

Please be mindful of our release process. Do not break it.